### PR TITLE
refactored code to be more stable / maintable

### PR DESCRIPTION
**Pull request by thmeitz**
Hi Robin,

i've begun to refactor your code. This is work in progress, so please don't merge it until I'm finished my work. I'm tested it  by unit test only and not against a test machine!

I've also made a Github pipeline. Please sign up for a free codecov account and add it as action in the pipeline. So we can check the code coverage of the project.

Please review and comment my PR.

I've written a bunch of unit tests. Have a look, what I've finished so far.
I'm working on Linux, so there is a Makefile you can run the tests. 

Furthermore we should reuse the http.Client: [connection re use in golang with http client](https://stuartleeks.com/posts/connection-re-use-in-golang-with-http-client/). I think we should adapt the code from [Zalando Skipper](https://github.com/zalando/skipper/blob/master/net/httpclient.go). The opensourced the code under the Apache 2 License. They had instrumentations for Jaeger etc. also.

At this time, we can't mock http requests, so we can't unit test the code. 

Youre written tests are not really integration tests, since you are not check the results from ksqldb. This is on my agenda also...

Best
Thomas

### CODE REVIEW

File: push.go [hunk 1]
```diff
@@ -89,21 +90,24 @@ func (cl *Client) Push(ctx context.Context, q string, rc chan<- Row, hc chan<- H
...
 			}
 
 			res, err := client.Do(req)
 			if err != nil {
-				return fmt.Errorf("Failed to execute HTTP request to cancel query\n%v", err)
+				return fmt.Errorf("failed to execute http request to cancel query\n%w", err)
 			}
+			defer res.Body.Close()
+
 			if res.StatusCode != 200 {
```
Comment (thmeitz) [inline push.go]: this should be also http.StatusOk

**Comment by rmoff** [comment]
This is great, thank you @thmeitz. I will have a look at the codecov/pipeline thing. 

There is an open PR #18 from @sebltm - is it OK to merge that or will it clash with what you're doing?

**Comment by rmoff** [comment]
> I've also made a Github pipeline. Please sign up for a free codecov account and add it as action in the pipeline. So we can check the code coverage of the project.

@thmeitz do you have a reference page for what to do here? I've created a codecov account but not clear what to do next :)  (FTR - I don't understand what codecov is to do either?…)

**Comment by thmeitz** [comment]
> This is great, thank you @thmeitz. I will have a look at the codecov/pipeline thing.
> 
> There is an open PR #18 from @sebltm - is it OK to merge that or will it clash with what you're doing?

Yes it's ok to merge the PR. If it's clashes, I'll fix it ;)

**Comment by thmeitz** [comment]
> > I've also made a Github pipeline. Please sign up for a free codecov account and add it as action in the pipeline. So we can check the code coverage of the project.
> 
> @thmeitz do you have a reference page for what to do here? I've created a codecov account but not clear what to do next :) (FTR - I don't understand what codecov is to do either?…)

- [Girhub codecov action](https://github.com/codecov/codecov-action)
- [About code coverage](https://docs.codecov.com/docs/about-code-coverage)

If you have problems, I'll send you a short howto.

When we develop software, hopefully we have unit tests (and other tests) to check all edge cases of our software development. There are a number of coverage criteria. Golang has code coverage tools that we can use to get code coverage. We upload these coverages to codecov and can use a Github action to check the new code coverage of a PR. Coverage is measured as a percentage. If we have a coverage of 20% in the current master, we can introduce a rule that the coverage must not change below the current coverage. If the coverage falls below the 20%, the PR cannot be merged into the master.

Since I've have generated the KSqlParser with  Antlr4 which has generated about 20k lines of code, we will have a bad code coverage for the parser :D

**Comment by thmeitz** [comment]
Have canceled the PR in consultation with Robin.
If you are interested in the new stuff, use my Github repo.
